rtl/eclrtl/rtldynfield.cpp

```diff
@@ -510,3 +510,481 @@ extern ECLRTL_API int getFieldNum(const char *fieldName, IOutputMetaData &  meta
+};
+
+inline constexpr FieldMatchType operator|(FieldMatchType a, FieldMatchType b) { return (FieldMatchType)((int)a | (int)b); }
+inline FieldMatchType &operator|=(FieldMatchType &a, FieldMatchType b) { return (FieldMatchType &) ((int &)a |= (int)b); }
+
+class GeneralRecordTranslator
+{
+public:
+    GeneralRecordTranslator(const RtlRecord &_destRecInfo, const RtlRecord &_srcRecInfo)
+    : destRecInfo(_destRecInfo), sourceRecInfo(_srcRecInfo)
+    {
+        matchInfo = new MatchInfo[destRecInfo.getNumFields()];
+        createMatchInfo();
+    }
+    ~GeneralRecordTranslator()
+    {
+        delete [] matchInfo;
+    }
+    void describe(unsigned indent = 0) const
+    {
+        for (int idx = 0; idx <  destRecInfo.getNumFields(); idx++)
+        {
+            const RtlFieldInfo *field = destRecInfo.queryField(idx);
+            const MatchInfo &match = matchInfo[idx];
+            if (match.matchType == match_none)
+                DBGLOG("%*sNo match for field %s", indent, "", field->name);
+            else
+            {
+                DBGLOG("%*sMatch (%d) for field %d for field %s", indent, "", match.matchType, match.matchIdx, field->name);
+                if (match.subTrans)
+                    match.subTrans->describe(indent+2);
+            }
+        }
+        if (!canTranslate())
+            DBGLOG("Translation is NOT possible");
+        else if (needsTranslate())
+            DBGLOG("Translation is possible");
+        else
+            DBGLOG("Translation is not necessary");
+    }
+    size32_t translate(ARowBuilder &builder, size32_t offset, const byte *sourceRec) const
+    {
+        unsigned numOffsets = sourceRecInfo.getNumVarFields() + 1;
+        size_t variableOffsets[numOffsets];
+        RtlRow sourceRow(sourceRecInfo, sourceRec, numOffsets, variableOffsets);
+        for (int idx = 0; idx < destRecInfo.getNumFields(); idx++)
+        {
+            //printf("before idx %d offset %d\n", idx, offset);
+            const RtlFieldInfo *field = destRecInfo.queryField(idx);
+            const RtlTypeInfo *type = field->type;
+            const MatchInfo &match = matchInfo[idx];
+            if (match.matchType == match_none || match.matchType==match_fail)
+                offset = type->buildNull(builder, offset, field);
+            else
+            {
+                unsigned matchField = match.matchIdx;
+                const RtlTypeInfo *sourceType = sourceRecInfo.queryType(matchField);
+                size_t sourceOffset = sourceRow.getOffset(matchField);
+                const byte *source = sourceRec + sourceOffset;
+                size_t copySize = sourceRow.getOffset(matchField+1) - sourceOffset;
+                switch (match.matchType)
+                {
+                case match_perfect:
+                {
+                    // Look ahead for other perfect matches and combine the copies
+                    while (idx < destRecInfo.getNumFields()-1)
```

Comment:
I think it would be worth revisiting later.  I am particularly thinking of the case where a single field is added to the payload and is being projected out.